tendermint/src/abci/data.rs

```diff
@@ -1,64 +1,78 @@
-use crate::{Error, Kind};
-use serde::{Deserialize, Deserializer, Serialize, Serializer};
-use std::{
-    fmt::{self, Display},
-    str::FromStr,
-};
-use subtle_encoding::hex;
+use crate::Error;
+use serde::{Deserialize, Serialize};
+use std::convert::{TryFrom, TryInto};
+use tendermint_proto::serializers::bytes::Base64;
 
 /// ABCI transaction data.
 ///
 /// Transactions are opaque binary blobs which are validated according to
 /// application-specific rules.
-#[derive(Clone, Debug, Eq, PartialEq, Default)]
+#[derive(Clone, Debug, Eq, PartialEq, Default, Serialize, Deserialize)]
+#[serde(try_from = "Base64", into = "Base64")]
 pub struct Data(Vec<u8>);
 
-impl Data {
-    /// Borrow the data as bytes
-    pub fn as_bytes(&self) -> &[u8] {
-        self.0.as_ref()
-    }
-}
+impl TryFrom<Vec<u8>> for Data {
+    type Error = Error;
 
-impl AsRef<[u8]> for Data {
-    fn as_ref(&self) -> &[u8] {
-        self.0.as_ref()
+    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
+        // Todo: Are there any restrictions on the incoming data?
```

Comment:
```suggestion
```

Doesn't look like it. Looks like, when it's present, it's just an [arbitrary string of bytes](https://github.com/tendermint/tendermint/blob/a8d86003085d88a6a78bb4f963298ef25b6acbd3/types/results_test.go#L12).

As such I'd recommend replacing this `TryFrom<Vec<u8>>` to be a `From<Vec<u8>>`.

Reply:
done